src/utils/Client.util.ts

```diff
@@ -274,3 +295,79 @@ export function getClientOptions(
     maxBodySizeMb: clientoptions?.maxBodySizeMb,
   };
 }
+
+/**
+ * Scheme, fqdn and port
+ */
+type EndpointTuple = [string, string, number];
+
+/**
+ * Parses an endpoint to scheme, fqdn and port
+ * Examples:
+ *  - http://localhost:3500 -> [http, localhost, 3500]
+ *  - localhost:3500 -> [http, localhost, 3500]
+ *  - :3500 -> [http, localhost, 3500]
+ *  - localhost -> [http, localhost, 80]
+ *  - https://localhost:3500 -> [https, localhost, 3500]
+ *  - [::1]:3500 -> [http, ::1, 3500]
+ *  - [::1] -> [http, ::1, 80]
+ *  - http://[2001:db8:1f70::999:de8:7648:6e8]:5000 -> [http, 2001:db8:1f70::999:de8:7648:6e8, 5000]
+ * @throws Error if the port is not a number
+ * @throws Error if the address is invalid
+ * @param address Endpoint address
+ * @returns EndpointTuple (scheme, fqdn, port)
+ */
+export function parseEndpoint(address: string): EndpointTuple {
```

Comment:
Is there a reason the native Node Lib https://nodejs.org/api/url.html is not being used?

Reply:
I tried different regex expressions and url parsing libraries when I was implementing the same thing in Python and to my surprise couldn't find anything that would work for what we need here. I didn't repeat the search for Node, but I can easily see if the library you suggested works by running the existing tests against it. Fingers crossed! :) I'll get back to you in a bit.

Reply:
Ok, this is what I got with the library you suggested. It's good for most of our cases, but it fails for many important ones.

I applied had to apply this little processing of results:

```js
  parsedScheme = myURL.protocol.replace(":", "");
  parsedHost = myURL.hostname.replace("[", "");
  parsedHost = parsedHost.replace("]", "");
  parsedPort = myURL.port || (myURL.protocol == "https:" ? 443 : 80);
```

| Endpoint                                                                                   | Scheme                                  | Hostname                                | Port |   Works for us    |
| ------------------------------------------------------------------------------------------ | --------------------------------------- | --------------------------------------- | ---- | ----- |
| [http://localhost](http://localhost)                                                       | http                                    | localhost                               | 80   | TRUE  |
| [http://localhost/v1/dapr](http://localhost/v1/dapr)                                       | http                                    | localhost                               | 80   | TRUE  |
| [http://localhost:5000](http://localhost:5000)                                             | http                                    | localhost                               | 5000 | TRUE  |
| [http://localhost:5000/v1/dapr](http://localhost:5000/v1/dapr)                             | http                                    | localhost                               | 5000 | TRUE  |
| [https://localhost](https://localhost)                                                     | https                                   | localhost                               | 443  | TRUE  |
| [https://localhost/v1/dapr](https://localhost/v1/dapr)                                     | https                                   | localhost                               | 443  | TRUE  |
| [https://localhost:5000](https://localhost:5000)                                           | https                                   | localhost                               | 5000 | TRUE  |
| [https://localhost:5000/v1/dapr](https://localhost:5000/v1/dapr)                           | https                                   | localhost                               | 5000 | TRUE  |
| 127.0.0.1                                                                                  | ERR_INVALID_URL                         |                                         |      | FALSE |
| 127.0.0.1/v1/dapr                                                                          | ERR_INVALID_URL                         |                                         |      | FALSE |
| 127.0.0.1:5000                                                                             | ERR_INVALID_URL                         |                                         |      | FALSE |
| 127.0.0.1:5000/v1/dapr                                                                     | ERR_INVALID_URL                         |                                         |      | FALSE |
| [http://127.0.0.1](http://127.0.0.1)                                                       | http                                    | 127.0.0.1                               | 80   | TRUE  |
| [http://127.0.0.1/v1/dapr](http://127.0.0.1/v1/dapr)                                       | http                                    | 127.0.0.1                               | 80   | TRUE  |
| [http://127.0.0.1:5000](http://127.0.0.1:5000)                                             | http                                    | 127.0.0.1                               | 5000 | TRUE  |
| [http://127.0.0.1:5000/v1/dapr](http://127.0.0.1:5000/v1/dapr)                             | http                                    | 127.0.0.1                               | 5000 | TRUE  |
| [https://127.0.0.1](https://127.0.0.1)                                                     | https                                   | 127.0.0.1                               | 443  | TRUE  |
| [https://127.0.0.1/v1/dapr](https://127.0.0.1/v1/dapr)                                     | https                                   | 127.0.0.1                               | 443  | TRUE  |
| [https://127.0.0.1:5000](https://127.0.0.1:5000)                                           | https                                   | 127.0.0.1                               | 5000 | TRUE  |
| [https://127.0.0.1:5000/v1/dapr](https://127.0.0.1:5000/v1/dapr)                           | https                                   | 127.0.0.1                               | 5000 | TRUE  |
| [2001:db8:1f70:0:999:de8:7648:6e8]                                                         | ERR_INVALID_URL                         |                                         |      | FALSE |
| [2001:db8:1f70:0:999:de8:7648:6e8]/v1/dapr                                                 | ERR_INVALID_URL                         |                                         |      | FALSE |
| [2001:db8:1f70:0:999:de8:7648:6e8]:5000                                                    | ERR_INVALID_URL                         |                                         |      | FALSE |
| [2001:db8:1f70:0:999:de8:7648:6e8]:5000/v1/dapr                                            | ERR_INVALID_URL                         |                                         |      | FALSE |
| http://[2001:db8:1f70:0:999:de8:7648:6e8]                                                  | http                                    | 2001:db8:1f70:0:999:de8:7648:6e8        | 80   | TRUE  |
| http://[2001:db8:1f70:0:999:de8:7648:6e8]/v1/dapr                                          | http                                    | 2001:db8:1f70:0:999:de8:7648:6e8        | 80   | TRUE  |
| http://[2001:db8:1f70:0:999:de8:7648:6e8]:5000                                             | http                                    | 2001:db8:1f70:0:999:de8:7648:6e8        | 5000 | TRUE  |
| http://[2001:db8:1f70:0:999:de8:7648:6e8]:5000/v1/dapr                                     | http                                    | 2001:db8:1f70:0:999:de8:7648:6e8        | 5000 | TRUE  |
| https://[2001:db8:1f70:0:999:de8:7648:6e8]                                                 | https                                   | 2001:db8:1f70:0:999:de8:7648:6e8        | 443  | TRUE  |
| https://[2001:db8:1f70:0:999:de8:7648:6e8]/v1/dapr                                         | https                                   | 2001:db8:1f70:0:999:de8:7648:6e8        | 443  | TRUE  |
| https://[2001:db8:1f70:0:999:de8:7648:6e8]:5000                                            | https                                   | 2001:db8:1f70:0:999:de8:7648:6e8        | 5000 | TRUE  |
| https://[2001:db8:1f70:0:999:de8:7648:6e8]:5000/v1/dapr                                    | https                                   | 2001:db8:1f70:0:999:de8:7648:6e8        | 5000 | TRUE  |
| [domain.com](http://domain.com)                                                            | ERR_INVALID_URL                         |                                         |      | FALSE |
| [domain.com/v1/grpc](http://domain.com/v1/grpc)                                            | ERR_INVALID_URL                         |                                         |      | FALSE |
| [domain.com:5000](http://domain.com:5000)                                                  | [domain.com](http://domain.com)         |                                         | 80   | FALSE |
| [domain.com:5000/v1/dapr](http://domain.com:5000/v1/dapr)                                  | [domain.com](http://domain.com)         |                                         | 80   | FALSE |
| [http://domain.com](http://domain.com)                                                     | http                                    | [domain.com](http://domain.com)         | 80   | TRUE  |
| [http://domain.com/v1/dapr](http://domain.com/v1/dapr)                                     | http                                    | [domain.com](http://domain.com)         | 80   | TRUE  |
| [http://domain.com:5000](http://domain.com:5000)                                           | http                                    | [domain.com](http://domain.com)         | 5000 | TRUE  |
| [http://domain.com:5000/v1/dapr](http://domain.com:5000/v1/dapr)                           | http                                    | [domain.com](http://domain.com)         | 5000 | TRUE  |
| [https://domain.com](https://domain.com)                                                   | https                                   | [domain.com](http://domain.com)         | 443  | TRUE  |
| [https://domain.com/v1/dapr](https://domain.com/v1/dapr)                                   | https                                   | [domain.com](http://domain.com)         | 443  | TRUE  |
| [https://domain.com:5000](https://domain.com:5000)                                         | https                                   | [domain.com](http://domain.com)         | 5000 | TRUE  |
| [https://domain.com:5000/v1/dapr](https://domain.com:5000/v1/dapr)                         | https                                   | [domain.com](http://domain.com)         | 5000 | TRUE  |
| [abc.domain.com](http://abc.domain.com)                                                    | ERR_INVALID_URL                         |                                         |      | FALSE |
| [abc.domain.com/v1/grpc](http://abc.domain.com/v1/grpc)                                    | ERR_INVALID_URL                         |                                         |      | FALSE |
| [abc.domain.com:5000](http://abc.domain.com:5000)                                          | [abc.domain.com](http://abc.domain.com) |                                         | 80   | FALSE |
| [abc.domain.com:5000/v1/dapr](http://abc.domain.com:5000/v1/dapr)                          | [abc.domain.com](http://abc.domain.com) |                                         | 80   | FALSE |
| [http://abc.domain.com/v1/dapr](http://abc.domain.com/v1/dapr)                             | http                                    | [abc.domain.com](http://abc.domain.com) | 80   | TRUE  |
| [http://abc.domain.com/v1/dapr](http://abc.domain.com/v1/dapr)                             | http                                    | [abc.domain.com](http://abc.domain.com) | 80   | TRUE  |
| [http://abc.domain.com:5000/v1/dapr](http://abc.domain.com:5000/v1/dapr)                   | http                                    | [abc.domain.com](http://abc.domain.com) | 5000 | TRUE  |
| [http://abc.domain.com:5000/v1/dapr/v1/dapr](http://abc.domain.com:5000/v1/dapr/v1/dapr)   | http                                    | [abc.domain.com](http://abc.domain.com) | 5000 | TRUE  |
| [https://abc.domain.com/v1/dapr](https://abc.domain.com/v1/dapr)                           | https                                   | [abc.domain.com](http://abc.domain.com) | 443  | TRUE  |
| [https://abc.domain.com/v1/dapr](https://abc.domain.com/v1/dapr)                           | https                                   | [abc.domain.com](http://abc.domain.com) | 443  | TRUE  |
| [https://abc.domain.com:5000/v1/dapr](https://abc.domain.com:5000/v1/dapr)                 | https                                   | [abc.domain.com](http://abc.domain.com) | 5000 | TRUE  |
| [https://abc.domain.com:5000/v1/dapr/v1/dapr](https://abc.domain.com:5000/v1/dapr/v1/dapr) | https                                   | [abc.domain.com](http://abc.domain.com) | 5000 | TRUE  |

Reply:
Can't those be resolved by just checking if the protocol exists, if it doesn't exist append a default one?

Reply:
Ok, done! With some prefixing, as you suggested, I was able to make all the tests pass. I think this is a better and more stable solution than maintaining our own parsing logic. Thanks @XavierGeerinck .

Reply:
AMAZING! Thank you